qualtran/bloqs/state_preparation/controlled_state_preparation.py

```diff
@@ -0,0 +1,484 @@
+            provided.
+        rot_reg_bitsizes: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        rom_values: the tuple of values to be loaded in the rom, which correspond to the angle of
+            each rotation. In order to get the rom value that corresponds to a given angle use
+            RotationTree.angle_to_rom_value(angle, rot_reg_bitsizes).
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    selection_bitsizes: int
+    rot_reg_bitsizes: int
+    rom_values: Tuple
+
+    @property
+    def signature(self):
+        return Signature.build(
+            prepare_control=1,
+            selection=self.selection_bitsizes,
+            qubit=1,
+            rot_reg=self.rot_reg_bitsizes,
+            phase_gradient=self.rot_reg_bitsizes,
+        )
+
+    def build_composite_bloq(self, bb: BloqBuilder, **soqs: SoquetT) -> Dict[str, SoquetT]:
+        """Parameters:
+        * prepare_control
+        * selection (not necessary if selection_bitsizes == 0)
+        * qubit
+        * rot_reg
+        * phase_gradient
+        """
+        qrom = QROM(
+            [np.array(self.rom_values)],
+            selection_bitsizes=(self.selection_bitsizes,),
+            target_bitsizes=(self.rot_reg_bitsizes,),
+            num_controls=2,
+        )
+        # both prepare_control and qubit will control the QROM so that this acts as a control Z
+        qrom_control = bb.join(np.array([soqs["prepare_control"], soqs["qubit"]]))
+        # load angles in rot_reg (line 1 of eq (8) in [1])
+        qrom_control, soqs = self._apply_qrom(qrom, bb, qrom_control, soqs)
+        # phase kickback via phase_grad += rot_reg (line 2 of eq (8) in [1])
+        soqs["rot_reg"], soqs["phase_gradient"] = bb.add(
+            Add(bitsize=self.rot_reg_bitsizes), a=soqs["rot_reg"], b=soqs["phase_gradient"]
+        )
+        # uncompute angle load in rot_reg to disentangle it from selection register
+        # (line 1 of eq (8) in [1])
+        qrom_control, soqs = self._apply_qrom(qrom, bb, qrom_control, soqs)
+        separated = bb.split(qrom_control)
+        soqs["prepare_control"] = separated[0]
+        soqs["qubit"] = separated[1]
+        return soqs
+
+    def _apply_qrom(
+        self, qrom: QROM, bb: BloqBuilder, qrom_control: SoquetT, soqs: Dict[str, SoquetT]
+    ):
+        if self.selection_bitsizes != 0:
+            qrom_control, soqs["selection"], soqs["rot_reg"] = bb.add(
+                qrom, control=qrom_control, selection=soqs["selection"], target0_=soqs["rot_reg"]
+            )
+        else:
+            qrom_control, soqs["rot_reg"] = bb.add(
+                qrom, control=qrom_control, target0_=soqs["rot_reg"]
+            )
+        return qrom_control, soqs
+
+
+# @attrs.frozen
```

Comment:
Great idea! I suppose that accessing array items is way faster than the old recursive version. Also, is there any reason to use bit logic instead of `2*i` and `2*i + 1`? I would assume that the assembly produced ends up being the same.

Reply:
> Also, is there any reason to use bit logic instead of 2*i and 2*i + 1

Not really. Just old habits die hard :)